www/source/blog/2017-10-24-node-with-db.html.md

```diff
@@ -0,0 +1,348 @@
+
+var app = express();
+mongoose.connect("mongodb://localhost:27017/test");
+setUpPassport();
+```
+
+First, delete this line - since we define the database url through Habitat:
+
+```
+mongoose.connect("mongodb://localhost:27017/test");
+```
+
+Now, add some content right after this line:
+
+```
+var app = express();
+```
+
+Like so:
+
+**app.js**
+
+```
+var app = express();
+var nconf = require('nconf');
+const nconf_file = process.env.APP_CONFIG || './default-config.json';
+nconf.file({ file: nconf_file });
+
+console.log("Loading Configuration");
+
+console.log(nconf.get('mongo'));
+var dbConnectionString = 'mongodb://' + nconf.get('mongo')['host'] + ':' + nconf.get('mongo')['port'] + "/test'";
+console.log(dbConnectionString);
+
+console.log("Connecting to Database");
+
+var mongoose = require('mongoose');
+mongoose.connect(dbConnectionString, { useMongoClient: true });
+```
+
+Wow, that's a lot.  Let's break this down:
+
+```
+var nconf = require('nconf');
+```
+
+First, we use an npm module called [nconf](https://www.npmjs.com/package/nconf).  Nconf is a tool to define configuration variables, then use them within the node application.
+
+
+```
+const nconf_file = process.env.APP_CONFIG;
+nconf.file({ file: nconf_file });
+```
+
+Then we read a configuration file and load it to nconf. This brings us to the question - where do we get this file?
+
+Let's create a template for this config file at habitat/config/config.json
+
+**habitat/config/config.json**
+
+```
+{
+		"mongo": {
+			{{~#eachAlive bind.database.members as |member|}}
+			{{#if @first}}
+			"host" : "{{member.sys.ip}}",
+			"port"   : "{{member.cfg.port}}"
+			{{/if}}
+			{{~/eachAlive}}
+			}
+}
+```
+
+Why does it look like this? Storing data in JSON is tricky with Handlebars templating because:
+* Each element needs a comma except the last element
+* Duplicate keys or IDs at the same depth will cause errors
+
+We allow for multiple members of a MongoDB cluster - but only want the IP/Port information of one to connect to it. When defining this config, we want to write out one entry from the members array. It could be the first one or the last one. The important thing is to only write once.
+
+Now that we have this defined, we can use this config:
+
+```
+var dbConnectionString = 'mongodb://' + nconf.get('mongo')['host'] + ':' + nconf.get('mongo')['port'] + "/test'";
+```
+
+Here, we create a db connection string - pulling the ('mongo')['host'] and ('mongo')['port'] values from the configuration file we just loaded.
+
+In this case, we uses the default database included with all new installs of MongoDB - the test database - at the end of the db connection string.
+
+```
+var mongoose = require('mongoose');
+mongoose.connect(dbConnectionString, { useMongoClient: true });
+```
+
+Finally, we use the mongoose node module to connect to the database. Go ahead and save and close the file.
+
+Open up your package.json file and add nconf as a dependency:
```

Comment:
Here you probably ask them to run `npm install nconf --save`. This will install the the `nconf` module and then write the version to the `package.json`. My thought is that `nconf` will likely break this functionality so I wouldn't fear them installing whatever is latest.

Reply:
Good point!  Fixing!